File: 139-Word Break.py
# dp
# Runtime: 44 ms, faster than 70.03% of Python3 online submissions for Word Break.
# Memory Usage: 12.5 MB, less than 100.00% of Python3 online submissions for Word Break.
class Solution:
    def wordBreak(self, s: 'str', wordDict: 'List[str]') -> 'bool':
        dp = [False] * (len(s)+1)
        dp[0] = True
        for i in range(1, len(s)+1):
            for j in range(i):
                if dp[j] and s[j:i] in wordDict:
                    dp[i] = True
                    break
        return dp[-1]

File: Algorithms/C++/139-Word_Break.cpp
// Runtime: 8 ms, faster than 77.07% of C++ online submissions for Word Break.
// Memory Usage: 12 MB, less than 60.38% of C++ online submissions for Word Break.
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<bool> dp(s.size() + 1, false);
        dp[0] = true;
        for (int i = 0; i < s.size(); ++i) {
            for (int j = i; j >= 0; --j) {
                string temp = s.substr(j, i - j + 1);
                if (dp[j])
                    if (find(wordDict.begin(), wordDict.end(), temp) != wordDict.end()) {
                        dp[i + 1] = true;
                        break;
                    }
            }
        }
        return dp[s.size()];
    }
};
